### PR TITLE
Integrates authorizations for commands to the teams repo

### LINKED ISSUE #34: Implement authorization with the team repository
We want to centralize authorization in the [rust-lang/team](https://github.com/rust-lang/team) repository, so this bot should use it to detect which users are authorized to interact with it. This should be implemented before deploying the bot.

*Changes in rust-lang/team:*

1. Replace the existing [`discord = "username"`](https://github.com/rust-lang/team/blob/6eeeee38b36c7cb51457d1a4340d966d31aee909/src/schema.rs#L55) field with `discord-id = 1234567890123`.
2. Add `discord_ids` to `Permission` in [the static API](https://github.com/rust-lang/team/blob/6eeeee38b36c7cb51457d1a4340d966d31aee909/rust_team_data/src/v1.rs#L81-L85), to allow querying which Discord IDs implement a permission.
3. Implement a `crates_io_ops_bot.APP_NAME` [permission](https://github.com/rust-lang/team/blob/master/src/permissions.rs). We'll want it to work similarly to the bors permission, where the list of applications allowed is encoded in the `permissions!` macro, to allow strict checking during validation.

*Changes in the bot:*

Every time we need to check whether a user is allowed or not, the following URL needs to be queried:

```
https://team-api.infra.rust-lang.org/v1/permissions/crates_io_ops_bot.APP_NAME.json
```

This way, changes in the team repo are instantly relected in the bot without needing any restart.

cc @nellshamrell

### PR DESCRIPTION
**Pull request by nellshamrell**
Fixes #34 

Signed-off-by: Nell Shamrell <nellshamrell@gmail.com>

### CODE REVIEW

File: src/authorizations/users.rs [hunk 1]
```diff
@@ -1,31 +1,76 @@
...
+
+    let team_request = team_client.client.get(
+        &String::from("https://team-api.infra.rust-lang.org/v1/permissions/crates_io_ops_bot.staging_crates_io.json")
+    );
+
+    let team_response = team_request.send().and_then(|res| res.error_for_status())?;
+
+    let response_text = team_response.text().unwrap();
+
+    let team_json: TeamResponse = serde_json::from_str(&response_text).unwrap();
```
Comment (pietroalbini) [inline src/authorizations/users.rs:51]:
```suggestion
    let team_json: TeamResponse = team_request
        .send()?
        .error_for_status()?
        .json()?;
```

Also, note that the JSON decoding here could fail, for example if GitHub Pages returns a 5xx error.

File: src/authorizations/users.rs [hunk 2]
```diff
@@ -1,31 +1,76 @@
...
+
+fn discord_id_in_list(id: &str, team_response: TeamResponse) -> bool {
+    team_response
+        .discord_ids
+        .contains(&id.parse::<usize>().unwrap())
+}
+
+fn team_info() -> TeamResponse {
+    get_team_info().unwrap()
+}
```
Comment (pietroalbini) [inline src/authorizations/users.rs:38]: Let's avoid unwrapping here. What I would do is change the `is_authorized` function to return a `Result<bool, Error>`, and wrap the error returned by `get_team_info` to add some context like "failed to fetch authorized users list from the Team API".

**Comment by nellshamrell** [comment]
@pietroalbini changes made! ty!

File: src/authorizations/users.rs [hunk 3]
```diff
@@ -1,31 +1,92 @@
-use crate::config::Config;
+use reqwest::blocking::Client as reqwest_client;
```
Comment (pietroalbini) [inline src/authorizations/users.rs:1]: It's a bit weird to have a snake_case struct :)

File: src/authorizations/users.rs [hunk 4]
```diff
@@ -1,31 +1,92 @@
...
+    );
+
+    let team_response = team_request.send()?;
+
+    let team_response = match team_response.error_for_status() {
+        Ok(team_response) => team_response,
+        Err(err) => {
+            return Err(Box::new(err));
+        }
+    };
```
Comment (pietroalbini) [inline src/authorizations/users.rs:41]: The `?` operator doesn't work here?

File: src/authorizations/users.rs [hunk 5]
```diff
@@ -1,31 +1,92 @@
...
+
+pub fn is_authorized(id: &str) -> Result<bool, Box<dyn Error>> {
+    let authorization_info = team_info();
+
+    let authorization_info = match authorization_info {
+        Ok(authorization_info) => authorization_info,
+        Err(e) => {
+            return Err(e);
+        }
+    };
```
Comment (pietroalbini) [inline src/authorizations/users.rs:60]: Same, why is `?` not used here?

**Comment by nellshamrell** [comment]
@pietroalbini - changes made!

File: src/authorizations/users.rs [hunk 6]
```diff
@@ -1,31 +1,78 @@
...
+    let team_response = team_request.send().and_then(|res| res.error_for_status())?;
+
+    let team_json: TeamResponse = team_response.json()?;
+
+    Ok(team_json)
+}
+
+fn team_info() -> std::result::Result<TeamResponse, Box<dyn Error>> {
+    get_team_info()
+}
```
Comment (pietroalbini) [inline src/authorizations/users.rs:43]: I don't think there is much point for this function to exist anymore :D

File: src/authorizations/users.rs [hunk 7]
```diff
@@ -1,31 +1,78 @@
...
+
+fn get_team_info() -> Result<TeamResponse, Box<dyn Error>> {
+    let team_client = TeamClient::new();
+
+    let team_request = team_client.client.get(
+        &String::from("https://team-api.infra.rust-lang.org/v1/permissions/crates_io_ops_bot.staging_crates_io.json")
+    );
+
+    let team_response = team_request.send().and_then(|res| res.error_for_status())?;
```
Comment (pietroalbini) [inline src/authorizations/users.rs:34]:
```suggestion
    let team_response = team_request.send()?.error_for_status()?;
```

You can remove the `and_then` here.

**Pull request merged**